### PR TITLE
[Flow] Add patterns to lower tensor.reshape to flow.tensor.reshape

### PR DESCRIPTION
**Pull request by saienduri**

### CODE REVIEW

**Comment by saienduri** [comment]
working on one more optimization, will reopen once done

**Comment by benvanik** [comment]
You can leave PRs open and push more changes to reduce noise. Marking them as draft tells reviewers that you aren't ready for the full review.

RE this change: please describe why this is required now (it hasn't been until now) and add tests.

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: I don't think this will quite work. You are essentially getting the sizes of the result of the reshape, and then replacing that same result with the Flow reshape, so the result of the original reshape is still live. You need to get the tensor sizes from the 1D index tensor (2nd argument of tensor.reshape instead).

I would keep the unranked case unsupported for now and use a sequence of `tensor.extract` operations on the index tensor to get the target shape. In most cases this index tensor should come from a `tensor.from_elements` so I would expect most of that to fold away.

Comment (saienduri) [reply compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: Yeah, that's why I closed it actually. Wanted to make that change. It has been made now.

**Comment by saienduri** [comment]
We were working on adding a fallback case in torch-mlir for aten.view that lowers into tensor::reshape. This tensor op is not supported in IREE and needs to be converted to a flow.tensor.reshape operation. Therefore, this change. Although we are not proceeding with the tensor::reshape fallback in torch-mlir, we decided we can still add the IREE implementation just so that tensor::reshape is supported. But, maybe it's best to add this when we actually decide it's ok to lower to tensor.reshape because I know there are concerns with that from Ramiro.

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: Nit: IREE style is to add `{` `}` everywhere.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
You need to add a test for this. Its also easier to review if I can see an example of what the IR transformation does.

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (saienduri) [reply compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: done

**Comment by saienduri** [comment]
had to make some changes (realized that for flow tensor you only add to the result dims if it is dynamic after digging into the flow tensor reshape usage/definition). So, now it successfully lowers and here is the IR Dump if it helps: 
`/// -----// IR Dump After ConvertToFlow (iree-flow-convert-to-flow) //----- //
module {
  func.func @static_tensor_reshape(%arg0: tensor<2x4xf32>, %arg1: tensor<2xindex>) -> tensor<1x8xf32> {
    %0 = flow.tensor.reshape %arg0 : tensor<2x4xf32> -> tensor<1x8xf32>
    return %0 : tensor<1x8xf32>
  }
}


module {
  func.func @static_tensor_reshape(%arg0: tensor<2x4xf32>, %arg1: tensor<2xindex>) -> tensor<1x8xf32> {
    %0 = flow.tensor.reshape %arg0 : tensor<2x4xf32> -> tensor<1x8xf32>
    return %0 : tensor<1x8xf32>
  }
}


// -----
// -----// IR Dump After ConvertToFlow (iree-flow-convert-to-flow) //----- //
module {
  func.func @dynamic_tensor_reshape(%arg0: tensor<2x4xf32>, %arg1: tensor<2xindex>) -> tensor<?x?xf32> {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %0 = flow.tensor.load %arg1[%c0] : tensor<2xindex>
    %1 = flow.tensor.load %arg1[%c1] : tensor<2xindex>
    %2 = flow.tensor.reshape %arg0 : tensor<2x4xf32> -> tensor<?x?xf32>{%0, %1}
    return %2 : tensor<?x?xf32>
  }
}


module {
  func.func @dynamic_tensor_reshape(%arg0: tensor<2x4xf32>, %arg1: tensor<2xindex>) -> tensor<?x?xf32> {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %0 = flow.tensor.load %arg1[%c0] : tensor<2xindex>
    %1 = flow.tensor.load %arg1[%c1] : tensor<2xindex>
    %2 = flow.tensor.reshape %arg0 : tensor<2x4xf32> -> tensor<?x?xf32>{%0, %1}
    return %2 : tensor<?x?xf32>
  }
}


// -----
// -----// IR Dump After ConvertToFlow (iree-flow-convert-to-flow) //----- //
module {
  func.func @mix_dynamic_and_static_tensor_reshape(%arg0: tensor<2x4xf32>, %arg1: tensor<2xindex>) -> tensor<1x?xf32> {
    %c1 = arith.constant 1 : index
    %0 = flow.tensor.load %arg1[%c1] : tensor<2xindex>
    %1 = flow.tensor.reshape %arg0 : tensor<2x4xf32> -> tensor<1x?xf32>{%0}
    return %1 : tensor<1x?xf32>
  }
}


module {
  func.func @mix_dynamic_and_static_tensor_reshape(%arg0: tensor<2x4xf32>, %arg1: tensor<2xindex>) -> tensor<1x?xf32> {
    %c1 = arith.constant 1 : index
    %0 = flow.tensor.load %arg1[%c1] : tensor<2xindex>
    %1 = flow.tensor.reshape %arg0 : tensor<2x4xf32> -> tensor<1x?xf32>{%0}
    return %1 : tensor<1x?xf32>
  }
}


// -----
// -----// IR Dump After ConvertToFlow (iree-flow-convert-to-flow) //----- //
module {
  func.func @dynamic_input_and_output_tensor_reshape(%arg0: tensor<?x4xf32>, %arg1: tensor<2xindex>) -> tensor<1x?xf32> {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %dim = tensor.dim %arg0, %c0 : tensor<?x4xf32>
    %0 = flow.tensor.load %arg1[%c1] : tensor<2xindex>
    %1 = flow.tensor.reshape %arg0 : tensor<?x4xf32>{%dim} -> tensor<1x?xf32>{%0}
    return %1 : tensor<1x?xf32>
  }
}


module {
  func.func @dynamic_input_and_output_tensor_reshape(%arg0: tensor<?x4xf32>, %arg1: tensor<2xindex>) -> tensor<1x?xf32> {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %dim = tensor.dim %arg0, %c0 : tensor<?x4xf32>
    %0 = flow.tensor.load %arg1[%c1] : tensor<2xindex>
    %1 = flow.tensor.reshape %arg0 : tensor<?x4xf32>{%dim} -> tensor<1x?xf32>{%0}
    return %1 : tensor<1x?xf32>
  }
}


// -----
// -----// IR Dump After ConvertToFlow (iree-flow-convert-to-flow) //----- //
module {
  func.func @from_elements_test_reshape(%arg0: tensor<?x4xf32>, %arg1: index, %arg2: index) -> tensor<?x1xf32> {
    %c0 = arith.constant 0 : index
    %dim = tensor.dim %arg0, %c0 : tensor<?x4xf32>
    %0 = flow.tensor.reshape %arg0 : tensor<?x4xf32>{%dim} -> tensor<?x1xf32>{%arg1}
    return %0 : tensor<?x1xf32>
  }
}


module {
  func.func @from_elements_test_reshape(%arg0: tensor<?x4xf32>, %arg1: index, %arg2: index) -> tensor<?x1xf32> {
    %c0 = arith.constant 0 : index
    %dim = tensor.dim %arg0, %c0 : tensor<?x4xf32>
    %0 = flow.tensor.reshape %arg0 : tensor<?x4xf32>{%dim} -> tensor<?x1xf32>{%arg1}
    return %0 : tensor<?x1xf32>
  }
}`

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/test/reshape.mlir
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/test/reshape.mlir]: Can you add a test for dynamic -> dynamic reshape? `flow.tensor.reshape` takes the dynamic input sizes as well.

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: Nit: you shouldnt need the `llvm::`

**Review by MaheshRavishankar — APPROVED** [review]
Change looks good to me. One thing to verify though. The shape of the tensor should be created from `tensor.from_elements`. Can we make sure (and maybe add a lit test) to ensure that the `tensor.from_elements` gets folded away. (you probably will need to add `-canonicalize` to the RUN command).

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (saienduri) [reply compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: done

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: Nit: Run clang-format on source code.

**Comment by saienduri** [comment]
> Change looks good to me. One thing to verify though. The shape of the tensor should be created from `tensor.from_elements`. Can we make sure (and maybe add a lit test) to ensure that the `tensor.from_elements` gets folded away. (you probably will need to add `-canonicalize` to the RUN command).

Added in the tensor.from_elements test and have updated the IR dump with new tests in previous comment (works fine).

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/test/reshape.mlir
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/test/reshape.mlir]: Nit: Add new line at end of file.

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/test/reshape.mlir
Comment (saienduri) [reply compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/test/reshape.mlir]: test has been added

**Review by MaheshRavishankar — APPROVED** [review]
This looks good to me. Please address comments and land.

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: I think we prefer using return values over output parameters: they improve readability, and often provide the same or better performance.

We've been following mixed Google and LLVM style guide. For this part, we follow Google: https://google.github.io/styleguide/cppguide.html#Output_Parameters

For more context: https://abseil.io/tips/176

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: style nit: LLVM style guide states that we use `auto` when the type is obvious from the context (i.e., dyn_cast<...>). 

https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: Maybe just dyn_cast it to `RankedTensorType`, so we don't need to check `hasRank` here?

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]:
```suggestion
      // flow.reshape only takes dynamic dims for the result, source dims (ignore static dimensions)
```

**Review by hanhanW — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (saienduri) [reply compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: Need ShapedType to check if dim is ShapedType::kDynamic.

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (saienduri) [reply compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: ok i see, changed to return value

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: nit: rename to `getDynamicTensorSizes` to make it clearer it is specifically getting dynamic sizes.

**Review by qedawkins — APPROVED** [review]
LGTM. (Note that you don't need to get approval from all reviewers before landing, as long as you've given enough time for all interested parties to take a look :)

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: RankedTensorType has all the traits that ShapedType has. There are other ways to check if they are dynamic or not, and that is a prefer way. E.g., `type.isDynamicDim(idx)`, `ShapedType::isDynamic(size)`, etc.

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: let's use structured binding, it is more common in IREE/MLIR codebase.

```suggestion
  for (const auto [idx, size] : enumerate(stp.getShape())) {
```

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: What does `stp` mean? Can we rename it to `type`?

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (saienduri) [reply compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: Ah ok that makes sense. Made the change

File: compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp
Comment (saienduri) [reply compiler/src/iree/compiler/Dialect/Flow/Conversion/TensorToFlow/Patterns.cpp]: It meant Shaped Type, but changed it now

**Review by hanhanW — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
